cpp/src/arrow/filesystem/mockfs.cc

```diff
@@ -187,6 +184,11 @@ class Entry : public EntryBase {
   ARROW_DISALLOW_COPY_AND_ASSIGN(Entry);
 };
 
+void Directory::AssignEntry(const std::string& s, std::unique_ptr<Entry> entry) {
```

Comment:
I don't quite understand how does this fix the issue...

Reply:
The AssignEntry implementation is now placed after the full definition of Entry so it can access its structure. Previously, Entry was only forward declared, and as the error message indicated, the `sizeof` a forward declaration is 0, which triggered the static_assert.

I'm not sure this is the most appropriate fix, though.